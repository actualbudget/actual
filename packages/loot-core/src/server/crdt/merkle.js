// TODO: Ok, several problems:
//
// * If nothing matches between two merkle trees, we should fallback
// * to the last window instead the front one (use 0 instead of the
// * key)
//
// * Need to check to make sure if account exists when handling
// * transaction changes in syncing

export function getKeys(trie) {
  return Object.keys(trie).filter(x => x !== 'hash');
}

export function keyToTimestamp(key) {
  // 16 is the length of the base 3 value of the current time in
  // minutes. Ensure it's padded to create the full value
  let fullkey = key + '0'.repeat(16 - key.length);

  // Parse the base 3 representation
  return parseInt(fullkey, 3) * 1000 * 60;
}

export function insert(trie, timestamp) {
  let hash = timestamp.hash();
  let key = Number(Math.floor(timestamp.millis() / 1000 / 60)).toString(3);

  trie = Object.assign({}, trie, { hash: trie.hash ^ hash });
  return insertKey(trie, key, hash);
}

function insertKey(trie, key, hash) {
  if (key.length === 0) {
    return trie;
  }
  const c = key[0];
  const n = trie[c] || {};
  return Object.assign({}, trie, {
    [c]: Object.assign({}, n, insertKey(n, key.slice(1), hash), {
      hash: n.hash ^ hash,
    }),
  });
}

export function build(timestamps) {
  let trie = {};
  for (let timestamp of timestamps) {
    insert(trie, timestamp);
  }
  return trie;
}

export function diff(trie1, trie2) {
  if (trie1.hash === trie2.hash) {
    return null;
  }

  let node1 = trie1;
  let node2 = trie2;
  let k = '';

  // This loop will eventually stop when it traverses down to find
  // where the hashes differ, or otherwise when there are no leaves
  // left (this shouldn't happen, if that's the case the hash check at
  // the top of this function should pass)
  while (1) {
    let keyset = new Set([...getKeys(node1), ...getKeys(node2)]);
    let keys = [...keyset.values()];
    keys.sort();

    let diffkey = null;

    // Traverse down the trie through keys that aren't the same. We
    // traverse down the keys in order. Stop in two cases: either one
    // of the nodes doesn't have the key, or a different key isn't
    // found. For the former case, we have to that because pruning is
    // lossy. We don't know if we've pruned off a changed key so we
    // can't traverse down anymore. For the latter case, it means two
    // things: either we've hit the bottom of the tree, or the changed
    // key has been pruned off. In the latter case we have a "partial"
    // key and will fill the rest with 0s. Note that if multiple older
    // messages were added into one trie, it's possible we will
    // generate a time that only encompasses *some* of the those
    // messages. Pruning is lossy, and we traverse down the left-most
    // changed time that we know of, because of pruning it might take
    // multiple passes to sync up a trie.
    for (let i = 0; i < keys.length; i++) {
      let key = keys[i];

      if (!node1[key] || !node2[key]) {
        break;
      }

      let next1 = node1[key];
      let next2 = node2[key];
      if (next1.hash !== next2.hash) {
        diffkey = key;
        break;
      }
    }

    if (!diffkey) {
      return keyToTimestamp(k);
    }

    k += diffkey;
    node1 = node1[diffkey] || {};
    node2 = node2[diffkey] || {};
  }
}

export function prune(trie, n = 2) {
  // Do nothing if empty
  if (!trie.hash) {
    return trie;
  }

  let keys = getKeys(trie);
  keys.sort();

  let next = { hash: trie.hash };
  keys = keys.slice(-n).map(k => (next[k] = prune(trie[k], n)));

  return next;
}

export function debug(trie, k = '', indent = 0) {
  const str =
    ' '.repeat(indent) +
    (k !== '' ? `k: ${k} ` : '') +
    `hash: ${trie.hash || '(empty)'}\n`;
  return (
    str +
    getKeys(trie)
      .map(key => {
        return debug(trie[key], key, indent + 2);
      })
      .join('')
  );
}
