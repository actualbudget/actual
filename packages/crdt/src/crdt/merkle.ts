// TODO: Ok, several problems:
//
// * If nothing matches between two merkle trees, we should fallback
// * to the last window instead the front one (use 0 instead of the
// * key)
//
// * Need to check to make sure if account exists when handling
// * transaction changes in syncing

import { Timestamp } from './timestamp';

/**
 * Represents a node within a trinary radix trie.
 */
export type TrieNode = {
  '0'?: TrieNode;
  '1'?: TrieNode;
  '2'?: TrieNode;
  hash?: number;
};

type NumberTrieNodeKey = keyof Omit<TrieNode, 'hash'>;

export function emptyTrie(): TrieNode {
  return { hash: 0 };
}

function isNumberTrieNodeKey(input: string): input is NumberTrieNodeKey {
  return ['0', '1', '2'].includes(input);
}

export function getKeys(trie: TrieNode): NumberTrieNodeKey[] {
  return Object.keys(trie).filter(isNumberTrieNodeKey);
}

export function keyToTimestamp(key: string): number {
  // 16 is the length of the base 3 value of the current time in
  // minutes. Ensure it's padded to create the full value
  let fullkey = key + '0'.repeat(16 - key.length);

  // Parse the base 3 representation
  return parseInt(fullkey, 3) * 1000 * 60;
}

/**
 * Mutates `trie` to insert a node at `timestamp`
 */
export function insert(trie: TrieNode, timestamp: Timestamp) {
  let hash = timestamp.hash();
  let key = Number(Math.floor(timestamp.millis() / 1000 / 60)).toString(3);

  trie = Object.assign({}, trie, { hash: (trie.hash || 0) ^ hash });
  return insertKey(trie, key, hash);
}

function insertKey(trie: TrieNode, key: string, hash: number): TrieNode {
  if (key.length === 0) {
    return trie;
  }
  const c = key[0];
  const t = isNumberTrieNodeKey(c) ? trie[c] : undefined;
  const n = t || {};
  return Object.assign({}, trie, {
    [c]: Object.assign({}, n, insertKey(n, key.slice(1), hash), {
      hash: (n.hash || 0) ^ hash,
    }),
  });
}

export function build(timestamps: Timestamp[]) {
  let trie = emptyTrie();
  for (let timestamp of timestamps) {
    insert(trie, timestamp);
  }
  return trie;
}

export function diff(trie1: TrieNode, trie2: TrieNode): number | null {
  if (trie1.hash === trie2.hash) {
    return null;
  }

  let node1 = trie1;
  let node2 = trie2;
  let k = '';

  // This loop will eventually stop when it traverses down to find
  // where the hashes differ, or otherwise when there are no leaves
  // left (this shouldn't happen, if that's the case the hash check at
  // the top of this function should pass)
  while (1) {
    let keyset = new Set([...getKeys(node1), ...getKeys(node2)]);
    let keys = [...keyset.values()];
    keys.sort();

    let diffkey = null;

    // Traverse down the trie through keys that aren't the same. We
    // traverse down the keys in order. Stop in two cases: either one
    // of the nodes doesn't have the key, or a different key isn't
    // found. For the former case, we have to that because pruning is
    // lossy. We don't know if we've pruned off a changed key so we
    // can't traverse down anymore. For the latter case, it means two
    // things: either we've hit the bottom of the tree, or the changed
    // key has been pruned off. In the latter case we have a "partial"
    // key and will fill the rest with 0s. Note that if multiple older
    // messages were added into one trie, it's possible we will
    // generate a time that only encompasses *some* of the those
    // messages. Pruning is lossy, and we traverse down the left-most
    // changed time that we know of, because of pruning it might take
    // multiple passes to sync up a trie.
    for (let i = 0; i < keys.length; i++) {
      let key = keys[i];

      let next1 = node1[key];
      let next2 = node2[key];

      if (!next1 || !next2) {
        break;
      }

      if (next1.hash !== next2.hash) {
        diffkey = key;
        break;
      }
    }

    if (!diffkey) {
      return keyToTimestamp(k);
    }

    k += diffkey;
    node1 = node1[diffkey] || emptyTrie();
    node2 = node2[diffkey] || emptyTrie();
  }

  return null;
}

export function prune(trie: TrieNode, n = 2): TrieNode {
  // Do nothing if empty
  if (!trie.hash) {
    return trie;
  }

  let keys = getKeys(trie);
  keys.sort();

  let next: TrieNode = { hash: trie.hash };

  // Prune child nodes.
  for (let k of keys.slice(-n)) {
    const node = trie[k];

    if (!node) {
      throw new Error(`TrieNode for key ${k} could not be found`);
    }

    next[k] = prune(node, n);
  }

  return next;
}

export function debug(trie: TrieNode, k = '', indent = 0): string {
  const str =
    ' '.repeat(indent) +
    (k !== '' ? `k: ${k} ` : '') +
    `hash: ${trie.hash || '(empty)'}\n`;
  return (
    str +
    getKeys(trie)
      .map(key => {
        const node = trie[key];
        if (!node) return '';
        return debug(node, key, indent + 2);
      })
      .join('')
  );
}
