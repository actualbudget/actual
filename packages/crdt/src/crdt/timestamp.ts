import murmurhash from 'murmurhash';
import { v4 as uuidv4 } from 'uuid';

import { TrieNode } from './merkle';

/**
 * Hybrid Unique Logical Clock (HULC) timestamp generator
 *
 * Globally-unique, monotonic timestamps are generated from the
 *    combination of the unreliable system time, a counter, and an
 *    identifier for the current node (instance, machine, process, etc.).
 *    These timestamps can accommodate clock stuttering (duplicate values),
 *    regression, and node differences within the configured maximum drift.
 *
 * In order to generate timestamps locally or for transmission to another
 *    node, use the send() method. For global causality, timestamps must
 *    be included in each message processed by the system. Whenever a
 *    message is received, its timestamp must be passed to the recv()
 *    method.
 *
 * Timestamps serialize into a 46-character collatable string
 *    example: 2015-04-24T22:23:42.123Z-1000-0123456789ABCDEF
 *    example: 2015-04-24T22:23:42.123Z-1000-A219E7A71CC18912
 *
 * The 64-bit hybrid clock is based on the HLC specification,
 * http://www.cse.buffalo.edu/tech-reports/2014-04.pdf
 */

export type Clock = {
  timestamp: MutableTimestamp;
  merkle: TrieNode;
};

// A mutable global clock
let clock: Clock;

export function setClock(clock_: Clock): void {
  clock = clock_;
}

export function getClock(): Clock {
  return clock;
}

export function makeClock(timestamp: Timestamp, merkle: TrieNode = {}) {
  return { timestamp: MutableTimestamp.from(timestamp), merkle };
}

export function serializeClock(clock: Clock): string {
  return JSON.stringify({
    timestamp: clock.timestamp.toString(),
    merkle: clock.merkle,
  });
}

export function deserializeClock(clock: string): Clock {
  let data;
  try {
    data = JSON.parse(clock);
  } catch (e) {
    data = {
      timestamp: '1970-01-01T00:00:00.000Z-0000-' + makeClientId(),
      merkle: {},
    };
  }

  const ts = Timestamp.parse(data.timestamp);

  if (!ts) {
    throw new Timestamp.InvalidError(data.timestamp);
  }

  return {
    timestamp: MutableTimestamp.from(ts),
    merkle: data.merkle,
  };
}

export function makeClientId() {
  return uuidv4().replace(/-/g, '').slice(-16);
}

let config = {
  // Allow 5 minutes of clock drift
  maxDrift: 5 * 60 * 1000,
};

const MAX_COUNTER = parseInt('0xFFFF');
const MAX_NODE_LENGTH = 16;

/**
 * timestamp instance class
 */
export class Timestamp {
  _state: { millis: number; counter: number; node: string };

  constructor(millis: number, counter: number, node: string) {
    this._state = {
      millis: millis,
      counter: counter,
      node: node,
    };
  }

  valueOf() {
    return this.toString();
  }

  toString() {
    return [
      new Date(this.millis()).toISOString(),
      ('0000' + this.counter().toString(16).toUpperCase()).slice(-4),
      ('0000000000000000' + this.node()).slice(-16),
    ].join('-');
  }

  millis() {
    return this._state.millis;
  }

  counter() {
    return this._state.counter;
  }

  node() {
    return this._state.node;
  }

  hash() {
    return murmurhash.v3(this.toString());
  }

  // Timestamp generator initialization
  // * sets the node ID to an arbitrary value
  // * useful for mocking/unit testing
  static init(options: { maxDrift?: number; node?: string } = {}) {
    if (options.maxDrift) {
      config.maxDrift = options.maxDrift;
    }

    setClock(
      makeClock(
        new Timestamp(
          0,
          0,
          options.node
            ? ('0000000000000000' + options.node).toString().slice(-16)
            : '',
        ),
      ),
    );
  }

  /**
   * maximum timestamp
   */
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  static max = Timestamp.parse(
    '9999-12-31T23:59:59.999Z-FFFF-FFFFFFFFFFFFFFFF',
  )!;

  /**
   * timestamp parsing
   * converts a fixed-length string timestamp to the structured value
   */
  static parse(timestamp: string | Timestamp): Timestamp | null {
    if (timestamp instanceof Timestamp) {
      return timestamp;
    }
    if (typeof timestamp === 'string') {
      let parts = timestamp.split('-');
      if (parts && parts.length === 5) {
        let millis = Date.parse(parts.slice(0, 3).join('-')).valueOf();
        let counter = parseInt(parts[3], 16);
        let node = parts[4];
        if (
          !isNaN(millis) &&
          millis >= 0 &&
          !isNaN(counter) &&
          counter <= MAX_COUNTER &&
          typeof node === 'string' &&
          node.length <= MAX_NODE_LENGTH
        ) {
          return new Timestamp(millis, counter, node);
        }
      }
    }
    return null;
  }

  /**
   * Timestamp send. Generates a unique, monotonic timestamp suitable
   * for transmission to another system in string format
   */
  static send(): Timestamp | null {
    if (!clock) {
      return null;
    }

    // retrieve the local wall time
    let phys = Date.now();

    // unpack the clock.timestamp logical time and counter
    let lOld = clock.timestamp.millis();
    let cOld = clock.timestamp.counter();

    // calculate the next logical time and counter
    // * ensure that the logical time never goes backward
    // * increment the counter if phys time does not advance
    let lNew = Math.max(lOld, phys);
    let cNew = lOld === lNew ? cOld + 1 : 0;

    // check the result for drift and counter overflow
    if (lNew - phys > config.maxDrift) {
      throw new Timestamp.ClockDriftError(lNew, phys, config.maxDrift);
    }
    if (cNew > MAX_COUNTER) {
      throw new Timestamp.OverflowError();
    }

    // repack the logical time/counter
    clock.timestamp.setMillis(lNew);
    clock.timestamp.setCounter(cNew);

    return new Timestamp(
      clock.timestamp.millis(),
      clock.timestamp.counter(),
      clock.timestamp.node(),
    );
  }

  // Timestamp receive. Parses and merges a timestamp from a remote
  // system with the local timeglobal uniqueness and monotonicity are
  // preserved
  static recv(msg: Timestamp): Timestamp | null {
    if (!clock) {
      return null;
    }

    // retrieve the local wall time
    let phys = Date.now();

    // unpack the message wall time/counter
    let lMsg = msg.millis();
    let cMsg = msg.counter();

    // assert the node id and remote clock drift
    // if (msg.node() === clock.timestamp.node()) {
    //   throw new Timestamp.DuplicateNodeError(clock.timestamp.node());
    // }
    if (lMsg - phys > config.maxDrift) {
      throw new Timestamp.ClockDriftError();
    }

    // unpack the clock.timestamp logical time and counter
    let lOld = clock.timestamp.millis();
    let cOld = clock.timestamp.counter();

    // calculate the next logical time and counter
    // . ensure that the logical time never goes backward
    // . if all logical clocks are equal, increment the max counter
    // . if max = old > message, increment local counter
    // . if max = messsage > old, increment message counter
    // . otherwise, clocks are monotonic, reset counter
    let lNew = Math.max(Math.max(lOld, phys), lMsg);
    let cNew =
      lNew === lOld && lNew === lMsg
        ? Math.max(cOld, cMsg) + 1
        : lNew === lOld
        ? cOld + 1
        : lNew === lMsg
        ? cMsg + 1
        : 0;

    // check the result for drift and counter overflow
    if (lNew - phys > config.maxDrift) {
      throw new Timestamp.ClockDriftError();
    }
    if (cNew > MAX_COUNTER) {
      throw new Timestamp.OverflowError();
    }

    // repack the logical time/counter
    clock.timestamp.setMillis(lNew);
    clock.timestamp.setCounter(cNew);

    return new Timestamp(
      clock.timestamp.millis(),
      clock.timestamp.counter(),
      clock.timestamp.node(),
    );
  }

  /**
   * zero/minimum timestamp
   */
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  static zero = Timestamp.parse(
    '1970-01-01T00:00:00.000Z-0000-0000000000000000',
  )!;

  static since = (isoString: string) => isoString + '-0000-0000000000000000';

  /**
   * error classes
   */
  static DuplicateNodeError = class DuplicateNodeError extends Error {
    constructor(node: string) {
      super('duplicate node identifier ' + node);
      this.name = 'DuplicateNodeError';
    }
  };

  static ClockDriftError = class ClockDriftError extends Error {
    constructor(...args: unknown[]) {
      super(
        ['maximum clock drift exceeded'].concat(args as string[]).join(' '),
      );
      this.name = 'ClockDriftError';
    }
  };

  static OverflowError = class OverflowError extends Error {
    constructor() {
      super('timestamp counter overflow');
      this.name = 'OverflowError';
    }
  };

  static InvalidError = class InvalidError extends Error {
    constructor(...args: unknown[]) {
      super(['timestamp is not valid'].concat(args.map(String)).join(' '));
      this.name = 'InvalidError';
    }
  };
}

class MutableTimestamp extends Timestamp {
  static from(timestamp: Timestamp) {
    return new MutableTimestamp(
      timestamp.millis(),
      timestamp.counter(),
      timestamp.node(),
    );
  }

  setMillis(n: number) {
    this._state.millis = n;
  }

  setCounter(n: number) {
    this._state.counter = n;
  }

  setNode(n: string) {
    this._state.node = n;
  }
}
