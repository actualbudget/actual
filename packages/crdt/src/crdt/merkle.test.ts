/* eslint-disable @typescript-eslint/no-non-null-assertion */
import * as merkle from './merkle';
import { Timestamp } from './timestamp';

function message(timestampStr: string, hash: number) {
  const timestamp = Timestamp.parse(timestampStr)!;
  timestamp.hash = () => hash;
  return { timestamp };
}

function insertMessages(
  trie: merkle.TrieNode,
  messages: ReturnType<typeof message>[],
) {
  messages.forEach(msg => {
    trie = merkle.insert(trie, msg.timestamp);
  });
  return trie;
}

describe('merkle trie', () => {
  test('adding an item works', () => {
    let trie = merkle.insert(
      merkle.emptyTrie(),
      Timestamp.parse('2018-11-12T13:21:40.122Z-0000-0123456789ABCDEF')!,
    );
    trie = merkle.insert(
      trie,
      Timestamp.parse('2018-11-13T13:21:40.122Z-0000-0123456789ABCDEF')!,
    );
    expect(trie).toMatchSnapshot();
  });

  test('diff returns the correct time difference', () => {
    let trie1 = merkle.emptyTrie();
    let trie2 = merkle.emptyTrie();

    const messages = [
      // First client messages
      message('2018-11-13T13:20:40.122Z-0000-0123456789ABCDEF', 1000),
      message('2018-11-14T13:05:35.122Z-0000-0123456789ABCDEF', 1100),
      message('2018-11-15T22:19:00.122Z-0000-0123456789ABCDEF', 1200),

      // Second client messages
      message('2018-11-20T13:19:40.122Z-0000-0123456789ABCDEF', 1300),
      message('2018-11-25T13:19:40.122Z-0000-0123456789ABCDEF', 1400),
    ];

    trie1 = merkle.insert(trie1, messages[0].timestamp);
    trie1 = merkle.insert(trie1, messages[1].timestamp);
    trie1 = merkle.insert(trie1, messages[2].timestamp);
    expect(trie1.hash).toBe(788);

    trie2 = merkle.insert(trie2, messages[3].timestamp);
    trie2 = merkle.insert(trie2, messages[4].timestamp);
    expect(trie2.hash).toBe(108);

    expect(new Date(merkle.diff(trie1, trie2)!).toISOString()).toBe(
      '2018-11-02T17:15:00.000Z',
    );

    trie1 = merkle.insert(trie1, messages[3].timestamp);
    trie1 = merkle.insert(trie1, messages[4].timestamp);
    trie2 = merkle.insert(trie2, messages[0].timestamp);
    trie2 = merkle.insert(trie2, messages[1].timestamp);
    trie2 = merkle.insert(trie2, messages[2].timestamp);
    expect(trie1.hash).toBe(888);
    expect(trie1.hash).toBe(trie2.hash);
  });

  test('diffing works with empty tries', () => {
    let trie1 = merkle.emptyTrie();
    let trie2 = merkle.insert(
      merkle.emptyTrie(),
      Timestamp.parse('2009-01-02T10:17:37.789Z-0000-0000testinguuid1')!,
    );

    expect(merkle.diff(trie1, trie2)).toBe(0);
  });

  test('pruning works and keeps correct hashes', () => {
    let messages = [
      message('2018-11-01T01:00:00.000Z-0000-0123456789ABCDEF', 1000),
      message('2018-11-01T01:09:00.000Z-0000-0123456789ABCDEF', 1100),
      message('2018-11-01T01:18:00.000Z-0000-0123456789ABCDEF', 1200),
      message('2018-11-01T01:27:00.000Z-0000-0123456789ABCDEF', 1300),
      message('2018-11-01T01:36:00.000Z-0000-0123456789ABCDEF', 1400),
      message('2018-11-01T01:45:00.000Z-0000-0123456789ABCDEF', 1500),
      message('2018-11-01T01:54:00.000Z-0000-0123456789ABCDEF', 1600),
      message('2018-11-01T02:03:00.000Z-0000-0123456789ABCDEF', 1700),
      message('2018-11-01T02:10:00.000Z-0000-0123456789ABCDEF', 1800),
      message('2018-11-01T02:19:00.000Z-0000-0123456789ABCDEF', 1900),
      message('2018-11-01T02:28:00.000Z-0000-0123456789ABCDEF', 2000),
      message('2018-11-01T02:37:00.000Z-0000-0123456789ABCDEF', 2100),
    ];

    let trie = merkle.emptyTrie();
    messages.forEach(msg => {
      trie = merkle.insert(trie, msg.timestamp);
    });
    expect(trie.hash).toBe(2496);
    expect(trie).toMatchSnapshot();

    let pruned = merkle.prune(trie);
    expect(pruned.hash).toBe(2496);
    expect(pruned).toMatchSnapshot();
  });

  test('diffing differently shaped tries returns correct time', () => {
    let messages = [
      message('2018-11-01T01:00:00.000Z-0000-0123456789ABCDEF', 1000),
      message('2018-11-01T01:09:00.000Z-0000-0123456789ABCDEF', 1100),
      message('2018-11-01T01:18:00.000Z-0000-0123456789ABCDEF', 1200),
      message('2018-11-01T01:27:00.000Z-0000-0123456789ABCDEF', 1300),
      message('2018-11-01T01:36:00.000Z-0000-0123456789ABCDEF', 1400),
      message('2018-11-01T01:45:00.000Z-0000-0123456789ABCDEF', 1500),
      message('2018-11-01T01:54:00.000Z-0000-0123456789ABCDEF', 1600),
      message('2018-11-01T02:03:00.000Z-0000-0123456789ABCDEF', 1700),
      message('2018-11-01T02:10:00.000Z-0000-0123456789ABCDEF', 1800),
      message('2018-11-01T02:19:00.000Z-0000-0123456789ABCDEF', 1900),
      message('2018-11-01T02:28:00.000Z-0000-0123456789ABCDEF', 2000),
      message('2018-11-01T02:37:00.000Z-0000-0123456789ABCDEF', 2100),
    ];

    let trie = insertMessages({}, messages);

    // Case 0: It always returns a base time when comparing with an
    // empty trie
    expect(new Date(merkle.diff(merkle.emptyTrie(), trie)!).toISOString()).toBe(
      '1970-01-01T00:00:00.000Z',
    );
    expect(new Date(merkle.diff(trie, merkle.emptyTrie())!).toISOString()).toBe(
      '1970-01-01T00:00:00.000Z',
    );

    // Case 1: Add an older message that modifies the trie in such a
    // way that it modifies the 1st out of 3 branches (so it will be
    // pruned away)
    let trie1 = insertMessages(trie, [
      message('2018-11-01T00:59:00.000Z-0000-0123456789ABCDEF', 900),
    ]);

    // Normal comparison works
    expect(new Date(merkle.diff(trie1, trie)!).toISOString()).toBe(
      '2018-11-01T00:54:00.000Z',
    );

    // Comparing the pruned new trie is lossy, so it returns an even older time
    expect(
      new Date(merkle.diff(merkle.prune(trie1), trie)!).toISOString(),
    ).toBe('2018-11-01T00:45:00.000Z');

    // Comparing the pruned original trie is just as lossy
    expect(
      new Date(merkle.diff(trie1, merkle.prune(trie))!).toISOString(),
    ).toBe('2018-11-01T00:45:00.000Z');

    // Pruning both tries is just as lossy as well, since the changed
    // key is pruned away in both cases and it won't find a changed
    // key so it bails at the point
    expect(
      new Date(
        merkle.diff(merkle.prune(trie1), merkle.prune(trie))!,
      ).toISOString(),
    ).toBe('2018-11-01T00:45:00.000Z');

    // Case 2: Add two messages similar to the above case, but the
    // second message modifies the 2nd key at the same level as the
    // first message modifying the 1st key
    let trie2 = insertMessages(trie, [
      message('2018-11-01T00:59:00.000Z-0000-0123456789ABCDEF', 900),
      message('2018-11-01T01:15:00.000Z-0000-0123456789ABCDEF', 1422),
    ]);

    // Normal comparison works
    expect(new Date(merkle.diff(trie2, trie)!).toISOString()).toBe(
      '2018-11-01T00:54:00.000Z',
    );

    // Same as case 1
    expect(
      new Date(merkle.diff(merkle.prune(trie2), trie)!).toISOString(),
    ).toBe('2018-11-01T00:45:00.000Z');

    // Same as case 1
    expect(
      new Date(merkle.diff(trie2, merkle.prune(trie))!).toISOString(),
    ).toBe('2018-11-01T00:45:00.000Z');

    // Pruning both tries is very lossy and this ends up returning a
    // time that only covers the second message. Syncing will need
    // multiple passes to sync up. This happens because the second
    // message provides a "changed path" that the diff takes which
    // ignores the first message.
    expect(
      new Date(
        merkle.diff(merkle.prune(trie2), merkle.prune(trie))!,
      ).toISOString(),
    ).toBe('2018-11-01T01:12:00.000Z');
  });
});
