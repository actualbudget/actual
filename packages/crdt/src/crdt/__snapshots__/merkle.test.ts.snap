// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merkle trie adding an item works 1`] = `
{
  "1": {
    "2": {
      "1": {
        "0": {
          "1": {
            "0": {
              "0": {
                "2": {
                  "0": {
                    "1": {
                      "1": {
                        "0": {
                          "2": {
                            "2": {
                              "0": {
                                "0": {
                                  "hash": 1983295247,
                                },
                                "hash": 1983295247,
                              },
                              "hash": 1983295247,
                            },
                            "hash": 1983295247,
                          },
                          "hash": 1983295247,
                        },
                        "hash": 1983295247,
                      },
                      "hash": 1983295247,
                    },
                    "hash": 1983295247,
                  },
                  "1": {
                    "0": {
                      "1": {
                        "0": {
                          "2": {
                            "0": {
                              "0": {
                                "0": {
                                  "hash": 1469038940,
                                },
                                "hash": 1469038940,
                              },
                              "hash": 1469038940,
                            },
                            "hash": 1469038940,
                          },
                          "hash": 1469038940,
                        },
                        "hash": 1469038940,
                      },
                      "hash": 1469038940,
                    },
                    "hash": 1469038940,
                  },
                  "hash": 565800531,
                },
                "hash": 565800531,
              },
              "hash": 565800531,
            },
            "hash": 565800531,
          },
          "hash": 565800531,
        },
        "hash": 565800531,
      },
      "hash": 565800531,
    },
    "hash": 565800531,
  },
  "hash": 565800531,
}
`;

exports[`merkle trie pruning works and keeps correct hashes 1`] = `
{
  "1": {
    "2": {
      "1": {
        "0": {
          "0": {
            "2": {
              "2": {
                "2": {
                  "1": {
                    "2": {
                      "2": {
                        "0": {
                          "0": {
                            "1": {
                              "2": {
                                "0": {
                                  "hash": 1000,
                                },
                                "hash": 1000,
                              },
                              "hash": 1000,
                            },
                            "2": {
                              "2": {
                                "0": {
                                  "hash": 1100,
                                },
                                "hash": 1100,
                              },
                              "hash": 1100,
                            },
                            "hash": 1956,
                          },
                          "1": {
                            "0": {
                              "2": {
                                "0": {
                                  "hash": 1200,
                                },
                                "hash": 1200,
                              },
                              "hash": 1200,
                            },
                            "1": {
                              "2": {
                                "0": {
                                  "hash": 1300,
                                },
                                "hash": 1300,
                              },
                              "hash": 1300,
                            },
                            "2": {
                              "2": {
                                "0": {
                                  "hash": 1400,
                                },
                                "hash": 1400,
                              },
                              "hash": 1400,
                            },
                            "hash": 1244,
                          },
                          "2": {
                            "0": {
                              "2": {
                                "0": {
                                  "hash": 1500,
                                },
                                "hash": 1500,
                              },
                              "hash": 1500,
                            },
                            "1": {
                              "2": {
                                "0": {
                                  "hash": 1600,
                                },
                                "hash": 1600,
                              },
                              "hash": 1600,
                            },
                            "2": {
                              "2": {
                                "0": {
                                  "hash": 1700,
                                },
                                "hash": 1700,
                              },
                              "hash": 1700,
                            },
                            "hash": 1336,
                          },
                          "hash": 1600,
                        },
                        "1": {
                          "0": {
                            "0": {
                              "1": {
                                "1": {
                                  "hash": 1800,
                                },
                                "hash": 1800,
                              },
                              "hash": 1800,
                            },
                            "1": {
                              "1": {
                                "1": {
                                  "hash": 1900,
                                },
                                "hash": 1900,
                              },
                              "hash": 1900,
                            },
                            "2": {
                              "1": {
                                "1": {
                                  "hash": 2000,
                                },
                                "hash": 2000,
                              },
                              "hash": 2000,
                            },
                            "hash": 1972,
                          },
                          "1": {
                            "0": {
                              "1": {
                                "1": {
                                  "hash": 2100,
                                },
                                "hash": 2100,
                              },
                              "hash": 2100,
                            },
                            "hash": 2100,
                          },
                          "hash": 3968,
                        },
                        "hash": 2496,
                      },
                      "hash": 2496,
                    },
                    "hash": 2496,
                  },
                  "hash": 2496,
                },
                "hash": 2496,
              },
              "hash": 2496,
            },
            "hash": 2496,
          },
          "hash": 2496,
        },
        "hash": 2496,
      },
      "hash": 2496,
    },
    "hash": 2496,
  },
  "hash": 2496,
}
`;

exports[`merkle trie pruning works and keeps correct hashes 2`] = `
{
  "1": {
    "2": {
      "1": {
        "0": {
          "0": {
            "2": {
              "2": {
                "2": {
                  "1": {
                    "2": {
                      "2": {
                        "0": {
                          "1": {
                            "1": {
                              "2": {
                                "0": {
                                  "hash": 1300,
                                },
                                "hash": 1300,
                              },
                              "hash": 1300,
                            },
                            "2": {
                              "2": {
                                "0": {
                                  "hash": 1400,
                                },
                                "hash": 1400,
                              },
                              "hash": 1400,
                            },
                            "hash": 1244,
                          },
                          "2": {
                            "1": {
                              "2": {
                                "0": {
                                  "hash": 1600,
                                },
                                "hash": 1600,
                              },
                              "hash": 1600,
                            },
                            "2": {
                              "2": {
                                "0": {
                                  "hash": 1700,
                                },
                                "hash": 1700,
                              },
                              "hash": 1700,
                            },
                            "hash": 1336,
                          },
                          "hash": 1600,
                        },
                        "1": {
                          "0": {
                            "1": {
                              "1": {
                                "1": {
                                  "hash": 1900,
                                },
                                "hash": 1900,
                              },
                              "hash": 1900,
                            },
                            "2": {
                              "1": {
                                "1": {
                                  "hash": 2000,
                                },
                                "hash": 2000,
                              },
                              "hash": 2000,
                            },
                            "hash": 1972,
                          },
                          "1": {
                            "0": {
                              "1": {
                                "1": {
                                  "hash": 2100,
                                },
                                "hash": 2100,
                              },
                              "hash": 2100,
                            },
                            "hash": 2100,
                          },
                          "hash": 3968,
                        },
                        "hash": 2496,
                      },
                      "hash": 2496,
                    },
                    "hash": 2496,
                  },
                  "hash": 2496,
                },
                "hash": 2496,
              },
              "hash": 2496,
            },
            "hash": 2496,
          },
          "hash": 2496,
        },
        "hash": 2496,
      },
      "hash": 2496,
    },
    "hash": 2496,
  },
  "hash": 2496,
}
`;
